src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1716,13 +1718,32 @@ internal void TraceCommandStart(NpgsqlConnectionStringBuilder settings)
     {
         Debug.Assert(CurrentActivity is null);
         if (NpgsqlActivitySource.IsEnabled)
-            CurrentActivity = NpgsqlActivitySource.CommandStart(settings, IsWrappedByBatch ? GetBatchFullCommandText() : CommandText, CommandType);
+        {
+            var tracingSettings = NpgsqlTracingOptions.Current;
+            var enableTracing = IsWrappedByBatch
+                ? tracingSettings?.FilterNpgsqlBatch?.Invoke(WrappingBatch!)
+                : tracingSettings?.FilterNpgsqlCommand?.Invoke(this);
+            if (enableTracing != false)
+            {
+                var spanName = IsWrappedByBatch
+                    ? tracingSettings?.ProvideSpanNameForNpgsqlBatch?.Invoke(WrappingBatch!)
+                    : tracingSettings?.ProvideSpanNameForNpgsqlCommand?.Invoke(this);
+                CurrentActivity = NpgsqlActivitySource.CommandStart(settings, IsWrappedByBatch ? GetBatchFullCommandText() : CommandText, CommandType, spanName);
+            }
```

Comment:
nit: possibly slightly clearer:

```suggestion
            var tracingSettings = NpgsqlTracingOptions.Current;

            (var enableTracing, string? spanName) = (true, null);
            if (tracingSettings is not null)
            {
                enableTracing = IsWrappedByBatch
                    ? tracingSettings.FilterNpgsqlBatch?.Invoke(WrappingBatch!) ?? true
                    : tracingSettings.FilterNpgsqlCommand?.Invoke(this) ?? true;

                spanName = IsWrappedByBatch
                    ? tracingSettings.ProvideSpanNameForNpgsqlBatch?.Invoke(WrappingBatch!)
                    : tracingSettings.ProvideSpanNameForNpgsqlCommand?.Invoke(this);
            }

            if (enableTracing)
            {
                CurrentActivity = NpgsqlActivitySource.CommandStart(
                    settings,
                    IsWrappedByBatch ? GetBatchFullCommandText() : CommandText,
                    CommandType,
                    spanName);
            }
```